src/main/java/uk/ac/ic/wlgitbridge/bridge/db/postgres/PostgresDBStore.java

```diff
@@ -0,0 +1,298 @@
+package uk.ac.ic.wlgitbridge.bridge.db.postgres;
+
+import uk.ac.ic.wlgitbridge.bridge.db.DBInitException;
+import uk.ac.ic.wlgitbridge.bridge.db.DBStore;
+import uk.ac.ic.wlgitbridge.bridge.db.ProjectState;
+import uk.ac.ic.wlgitbridge.util.Log;
+
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+
+public class PostgresDBStore implements DBStore {
+
+  private final Connection connection;
+
+  public PostgresDBStore(String connectionString, String user, String pass) {
+    Log.info("Initialize PostgresDBStore");
+    try {
+      Class.forName("org.postgresql.Driver");
+      connection = DriverManager
+        .getConnection(connectionString, user, pass);
+    } catch (Exception e) {
+      Log.error("Error connecting to Postgres: {}", e.getMessage());
+      throw new DBInitException(e);
+    }
+  }
+
+  public void dangerousResetDatabaseDontDoIt() {
+    try {
+      Statement statement = connection.createStatement();
+      statement.execute("delete from url_index_store;");
+      statement.execute("delete from projects;");
+      statement.close();
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public int getNumProjects() {
+    try {
+      Statement statement = connection.createStatement();
+      ResultSet rs = statement.executeQuery("SELECT count(*) from projects");
+      if (rs.next()) {
+        int result = rs.getInt(1);
+        rs.close();
+        statement.close();
+        return result;
+      } else {
+        throw new Exception("No results from query");
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public List<String> getProjectNames() {
```

Comment:
This seems like it would be impractical at scale, but it's in the base class and I guess we can use it for testing?

Reply:
Good point. I think this only gets called when the bridge is initialized:

```
repoStore.purgeNonexistentProjects(dbStore.getProjectNames());
```

Reply:
I see - I guess as things grow this looks like a somewhat backwards way of doing things from a performance perspective. We're currently:

* Retrieving a list of all projects in the database
* Removing the projects on disk that aren't in the list

The list from the database is always going to be bigger than the list of projects on disk, so it takes longer to compute the set from that way around.

But more importantly, there's a race condition waiting to happen here:

* Pod 1 is running and is about to get replaced with Pod 2
* Pod 2 wants to purge non-existent projects. It retrieves a list of projects from the database
* Pod 1 creates a new project on-disk
* Pod 2 deletes the project on disk (with `deleteInDirectoryApartFrom`) because it's not in the list

I think this will need to be replaced with something that checks in a more granular fashion.
```JavaScript
gitReposOnDisk = fs.glob(allTheProjects)
for (const repo of gitReposOnDisk) {
  if (!projectExistsInDatabase(repo)) {
    removeProjectFromDisk(repo)
  }
}
```

Thinking about it, this still potentially dangerous because `projectExistsInDatabase` would not lock a row in this case. So there is another chance of a race condition where:

* Pod 1 is waiting to be deleted
* Pod 2 comes up and starts deleting
* Pod 2 checks if a project exists in the database, that it found on disk
* Pod 1 creates the project in the databse
* Pod 2 deletes the project on disk

This could happen in two cases, depending on how things will be set up in the locking PR. -- I'm not sure what the order of events is here, in terms of when things are written to the disk. Will we add something to postgres *before* writing anything to the disk, or is the current behaviour to receive the data to disk and then add it to postgres? - If it's the latter, there is still a nontrivial chance of this race condition occurring.

If we touch the database *first*, then the race condition could only occur when something is on the disk but not in the database. This shouldn't occur during normal operation, assuming that the paragraph above is up to date, so would only occur if something has gone wrong or we've had to manually edit the db.

For full belt & braces, the logic should look something like
```JavaScript
gitReposOnDisk = fs.glob(allTheProjects)
for (const repo of gitReposOnDisk) {
  if (!projectExistsInDatabase(repo)) {
    const row = insertRowIntoPostgresAndLock(repo)
    removeProjectFromDisk(repo)
    deleteRowFromPostgresAndUnlock(row)
  }
}
```
which is faffy, but safer.

Reply:
I agree. However, this would require a bit of re-writing, and I think it would be better done after we have row-based locks :+1: